tools/build_utils.py

```diff
@@ -487,17 +487,43 @@ def install_ngraph_tf(tf_version, venv_dir, ngtf_pip_whl):
     print(ngraph_bridge.__version__)
 
 
+# Note: version can be a branch-name or tag or commit-id
 def download_repo(target_name, repo, version, submodule_update=False):
-    # First download to a temp folder
-    call(["git", "clone", repo, target_name])
+    if not os.path.exists(target_name):
+        call(["git", "clone", repo, target_name])
 
     pwd = os.getcwd()
-    os.chdir(target_name)
+    try:
+        os.chdir(target_name)
+    except:
+        raise Exception("Cannot find dir: " + target_name)
+
+    curr_remote = subprocess.check_output(
+        "git remote get-url origin", shell=True).rstrip()
+    if curr_remote != repo:
+        command_executor(["git", "remote set-url origin " + repo])
+    command_executor(["git", "fetch"])
 
-    # checkout the specified branch and get the latest changes
-    call(["git", "fetch"])
-    command_executor(["git", "checkout", version])
-    call(["git", "pull"])
+    is_branch = ""
+    try:
+        is_branch = subprocess.check_output(
+            "git show-ref refs/heads/" + version + " refs/remotes/origin/" +
+            version,
+            shell=True).rstrip()
+    except:
+        pass
+    if is_branch == "":
+        # we were given a tag/commit ref
+        curr_git_branch = subprocess.check_output(
+            "git branch | grep '*' | sed -e 's#.* ##'", shell=True).rstrip()
+        if curr_git_branch == "temp_branch":
+            call(["git", "checkout", "master"])
+        call(["git", "branch", "-D", "temp_branch"])
+        command_executor(["git", "checkout", "-b temp_branch", version])
```

Comment:
If you checkout a commit id / tag directly, it puts workspace into a "detached head" state. So, you won't be able to